compendium/DeclarativeServices/test/TestBundleValidation.cpp

```diff
@@ -0,0 +1,393 @@
+        receivedBundleValidationErrorEvent = true;
+      }
+
+      if (evt.GetType() ==
+            cppmicroservices::FrameworkEvent::Type::FRAMEWORK_ERROR &&
+          evt.GetBundle().GetSymbolicName() == "TestBundleDSTOI6") {
+        receivedSecondBundleValidationErrorEvent = true;
+      }
+    });
+
+  test::InstallAndStartDS(f.GetBundleContext());
+
+  auto sDSSvcRef = f.GetBundleContext()
+                     .GetServiceReference<cppmicroservices::service::component::
+                                            runtime::ServiceComponentRuntime>();
+  ASSERT_TRUE(sDSSvcRef);
+  auto dsRuntimeService =
+    f.GetBundleContext()
+      .GetService<
+        cppmicroservices::service::component::runtime::ServiceComponentRuntime>(
+        sDSSvcRef);
+  ASSERT_TRUE(dsRuntimeService);
+
+  test::InstallLib(f.GetBundleContext(), "TestBundleDSTOI1");
+  auto bundles = f.GetBundleContext().GetBundles();
+  auto bundleIter = std::find_if(
+    bundles.begin(), bundles.end(), [](const cppmicroservices::Bundle& b) {
+      return (b.GetSymbolicName() == "TestBundleDSTOI1");
+    });
+
+  ASSERT_THROW(bundleIter->Start(), cppmicroservices::SecurityException);
+  // a bundle validation function which returns false must cause the
+  // Framework not to start the bundle and it should not be loaded
+  // into the process.
+  EXPECT_EQ(bundleIter->GetState(),
+            cppmicroservices::Bundle::State::STATE_RESOLVED);
+  ASSERT_TRUE(receivedBundleValidationErrorEvent);
+
+  // test starting a delayed activation ds component
+  // in this case, starting the bundle does not cause the shared library to load
+  // the shared library is loaded on the first call to "GetService"
+  test::InstallLib(f.GetBundleContext(), "TestBundleDSTOI6");
+  bundles = f.GetBundleContext().GetBundles();
+  bundleIter = std::find_if(
+    bundles.begin(), bundles.end(), [](const cppmicroservices::Bundle& b) {
+      return (b.GetSymbolicName() == "TestBundleDSTOI6");
+    });
+
+  ASSERT_NO_THROW(bundleIter->Start());
+
+  struct Interface1Impl final : public test::Interface1
+  {
+    std::string Description() override { return "foo"; }
+  };
+  f.GetBundleContext().RegisterService<test::Interface1>(
+    std::make_shared<Interface1Impl>());
+
+  auto svcRef = f.GetBundleContext().GetServiceReference<test::Interface2>();
+  ASSERT_TRUE(svcRef);
+  ASSERT_THROW(auto svcObj = f.GetBundleContext().GetService(svcRef),
+               cppmicroservices::SecurityException);
+
+  // a bundle validation function which returns false must cause the
+  // service component not to be enabled
+  auto compDesc = dsRuntimeService->GetComponentDescriptionDTO(
+    *bundleIter, "sample::ServiceComponent6");
+  ASSERT_FALSE(dsRuntimeService->IsComponentEnabled(compDesc));
+  ASSERT_TRUE(receivedSecondBundleValidationErrorEvent);
+
+  f.GetBundleContext().RemoveListener(std::move(token));
+  f.Stop();
+  f.WaitForStop(std::chrono::milliseconds::zero());
+}
```

Comment:
Just wondering if it would be worthwhile to have one or more concurrent tests to see what happens when the validation function returns a failure or an exception when multiple threads are accessing it at the same time. Maybe one of the GetService failures since we know that's a common use case.